### ISSUE TITLE
NativeAOT Windows - System.InvalidCastException: Specified cast is not valid

### ISSUE THREAD

**Issue by e012345678** [dotnet/runtime#89672]
I have followed the guide here https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net7 to try compile our app with Native AOT on .NET 8 Preview 6. However, I am getting this error:

```
EXEC : error : Specified cast is not valid. [C:\git\MyApp.csp
roj::TargetFramework=net8.0-windows10.0.19041.0]
  System.InvalidCastException: Specified cast is not valid.
     at System.Runtime.TypeCast.CheckCastClass(MethodTable*, Object) + 0x31
     at ILCompiler.MetadataManager.IsReflectionBlocked(TypeDesc) + 0x132
     at ILCompiler.MetadataManager.IsReflectionBlocked(Instantiation) + 0x3d
     at ILCompiler.MetadataManager.IsReflectionBlocked(TypeDesc) + 0xfc
     at ILCompiler.UsageBasedMetadataManager.HasConditionalDependenciesDueToEETypePresence(TypeDesc) + 0xec
     at ILCompiler.DependencyAnalysis.EETypeNode..ctor(NodeFactory, TypeDesc) + 0x10a
     at ILCompiler.DependencyAnalysis.NodeFactory.CreateNecessaryTypeNode(TypeDesc) + 0xfb
     at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey) + 0x11
     at ILCompiler.DependencyAnalysis.NodeFactory.NecessaryTypeSymbol(TypeDesc) + 0xe7
     at ILCompiler.DependencyAnalysis.NonGCStaticsNode.GetConditionalStaticDependencies(NodeFactory) + 0x32
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1) + 0
  xf3
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1) + 0x2b
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack() + 0xb5
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() + 0x50
     at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan() + 0x14
     at ILCompiler.Program.<Run>g__RunScanner|4_0(Program.<>c__DisplayClass4_0&) + 0x192
     at ILCompiler.Program.Run() + 0x1f9b
     at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass215_0.<.ctor>b__0(InvocationContext context) + 0x2f8
```

How can I find out more information of what this means and how it can be fixed? The publish command I used is: dotnet publish -f:net8.0-windows10.0.19041.0 -r:win10-x64 -c:Release

**Comment by EgorBo** [comment]
@e012345678 is that app public?

**Comment by e012345678** [comment]
Hi @EgorBo thanks for the reply, unfortunately no.

**Comment by am11** [comment]
> dotnet publish -f:net8.0-windows10.0.19041.0 -r:win10-x64 -c:Release

The framework (-f) value has TPI, which suggests that your app is using CsWinRT. This issue was recently fixed downstream: https://github.com/dotnet/runtime/issues/68278#issuecomment-1635217677). Currently, CsWinRT does not fully support AOT, so even after its next release, chances are your app may not build or run. You can follow the progress in their issue tracker for AOT support: https://github.com/microsoft/CsWinRT/labels/AOT.

**Comment by e012345678** [comment]
Thanks for the reply @am11 . Why is the NativeAOT marketed as it supports Windows (on the link in my first post) but it does not due to "CsWinRT does not fully support AOT". What fine print/requirements am I missing?

**Comment by am11** [comment]
> What fine print/requirements am I missing?

If your project or one of its dependencies are relying on one of the limitations: https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net8plus#limitations-of-native-aot-deployment, publishing as AOT app will not work.

**Comment by e012345678** [comment]
Hi @am11, is there anyway I can find out which of my dependencies have this limitation? Basically "what part of my app do I need to remove?"

**Comment by MichalPetryka** [comment]
> is there anyway I can find out which of my dependencies have this limitation?

Putting those options in your csproj should enable all the analyzers that will log such info at build time:
```xml
        <PublishAot>true</PublishAot>
        <EnableAOTAnalyzer>true</EnableAOTAnalyzer>
        <EnableSingleFileAnalyzer>true</EnableSingleFileAnalyzer>
        <EnableTrimAnalyzer>true</EnableTrimAnalyzer>
        <SuppressTrimAnalysisWarnings>false</SuppressTrimAnalysisWarnings>
        <TrimmerSingleWarn>false</TrimmerSingleWarn>
```

**Comment by MichalStrehovsky** [comment]
> Putting those options in your csproj should enable all the analyzers that will log such info at build time:

PublishAot is enough to enable all the analyzers. The last two change analysis behaviors and should be set if one want such behaviors.

I'm closing this Won't fix with the same justification as https://github.com/dotnet/runtime/issues/68278#issuecomment-1608635180

**Comment by Sergio0694** [comment]
Getting the same build error with CsWinRT `2.0.4-prerelease.230731.4`, which should have the fix from https://github.com/microsoft/CsWinRT/pull/1340 (also mentioned in #68278). Can't build a blank WinUI 3 app (with no XAML) with NativeAOT at all. FYI @manodasanW 🤔

Can also open a new issue in the CsWinRT repo to track this there.

**Comment by manodasanW** [comment]
> Getting the same build error with CsWinRT `2.0.4-prerelease.230731.4`, which should have the fix from [microsoft/CsWinRT#1340](https://github.com/microsoft/CsWinRT/pull/1340) (also mentioned in #68278). Can't build a blank WinUI 3 app with NativeAOT at all. FYI @manodasanW 🤔
> 
> Can also open a new issue in the CsWinRT repo to track this there.

You will need an updated Windows SDK projection with that CsWinRT version.
